### PR TITLE
Support DeviceInfo on Web

### LINKED ISSUE #55: Plans to support web?
Are there any plans to support web? It should at least be documented that it does not currently support web. Just tried it out on a personal web project, but it is not working.

### PR DESCRIPTION
**Pull request by brianegan-wiredash**
- DeviceInfo requires dart:io, which is not allowed on web
  - Therefore, turn DeviceInfo into a data class
  - Export a dart:io and dart:html version
  - Dart:io continues to use Platform to get OS information
  - Dart:html now includes a UserAgentParser to parse browser name and version
  - Update serialization for FeedbackItem and PendingFeedbackItem. toJson is strictly for json serialization, no sub-strings of json allowed.
  - Add new toMultipartFormField method to FeedbackItem for interop with the api layer.
  - Create UserAgentParser to parse Browser Type and Version from the browser's user agent
  - Finalizes Flutter Web Compatibility!

### Questions

  - Right now, I'm sending through the browser name and version for the `platformOS` and `platformOSBuild`, and the full User Agent string for the "platformVersion". Should we adjust the DeviceInfo data model on the backend to include `Browser` and `Browser Version` for Flutter web projects as well, and use the User Agent to sniff the Operating System and Version? I think that might actually be the way to go.

Fixes #55

### CODE REVIEW

File: lib/src/common/device_info/base_device_info_generator.dart [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+  BaseDeviceInfoGenerator(
+    this.buildInfo,
+    this.window,
+  );
+
+  @override
+  DeviceInfo generate() {
+    return DeviceInfo(
+      appIsDebug: kDebugMode,
```
Comment (flschweiger) [inline lib/src/common/device_info/base_device_info_generator.dart:20]: All my life I didn't know about this constant 😔

File: lib/src/common/device_info/base_device_info_generator.dart [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+      appVersion: buildInfo.buildVersion,
+      buildNumber: buildInfo.buildNumber,
+      buildCommit: buildInfo.buildCommit,
+      deviceId: buildInfo.deviceId,
+      locale: window.locale.toString(),
```
Comment (flschweiger) [inline lib/src/common/device_info/base_device_info_generator.dart:25]: On macOS (and some other desktops?) the locale is / was null back when I implemented the first version of DeviceInfo. Do you know if that was fixed recently @brianegan-wiredash? Should we think about a possible fix here?

File: lib/src/feedback/feedback_model.dart [hunk 3]
```diff
@@ -86,7 +84,7 @@ class FeedbackModel with ChangeNotifier {
     loading = true;
 
     final item = FeedbackItem(
-      deviceInfo: json.encode(DeviceInfo.generate(_buildInfoManager)),
+      deviceInfo: _deviceInfoGenerator.generate(),
```
Comment (flschweiger) [inline lib/src/feedback/feedback_model.dart:87]: This is a way nicer solution 🙌

**Review by flschweiger — APPROVED** [review]
Really cool stuff, let's take Wiredash to the web 🚀

**Comment by ghenry** [comment]
On Tue, 10 Nov 2020, 19:31 Frederik Schweiger, <notifications@github.com>
wrote:

> *@flschweiger* approved this pull request.
>
> Really cool stuff, let's take Wiredash to the web 🚀
> ------------------------------
>
> In lib/src/common/device_info/base_device_info_generator.dart
> <https://github.com/wiredashio/wiredash-sdk/pull/98#discussion_r520812130>
> :
>
> > +import 'package:wiredash/src/common/device_info/device_info.dart';
>
> +import 'package:wiredash/src/common/device_info/device_info_generator.dart';
>
> +
>
> +abstract class BaseDeviceInfoGenerator implements DeviceInfoGenerator {
>
> +  final BuildInfoManager buildInfo;
>
> +  final ui.Window window;
>
> +
>
> +  BaseDeviceInfoGenerator(
>
> +    this.buildInfo,
>
> +    this.window,
>
> +  );
>
> +
>
> +  @override
>
> +  DeviceInfo generate() {
>
> +    return DeviceInfo(
>
> +      appIsDebug: kDebugMode,
>
>
> All my life I didn't know about this constant 😔
>

I didn't know either, but now everyone should! :)

https://github.com/flutter/website/issues/4976#issuecomment-725097962

File: lib/src/common/device_info/base_device_info_generator.dart (hunk 2 above)
Comment (brianegan-wiredash) [reply lib/src/common/device_info/base_device_info_generator.dart:25]: Ah, that's really important. I'll make sure to do a null check for that. Thanks!

File: test/common/device_info/device_info_test.dart [hunk 4]
```diff
@@ -0,0 +1,119 @@
+import 'package:test/test.dart';
+import 'package:wiredash/src/common/device_info/device_info.dart';
+
+void main() {
+  group('DeviceInfo', () {
```
Comment (passsy) [inline test/common/device_info/device_info_test.dart:5]: wanna add a test where you test the equals/hashcode method? 
```dart
final info = DeviceInfo(...);
final copy = DeviceInfo.fromJson(info.toJson());
expect(info, isEqualTo(copy));
expect(info.hashCode, isEqualTo(copy.hashCode));
```

File: lib/src/common/device_info/user_agent_parser.dart [hunk 5]
```diff
@@ -0,0 +1,421 @@
...
+
+  final String userAgent;
+
+  const UserAgentParser(this.userAgent);
+
+  String get browserName {
+    for (final matcher in _browserMatchers) {
+      for (final regex in matcher.regexes) {
+        if (regex.hasMatch(userAgent)) {
+          final name = matcher.name(regex.firstMatch(userAgent));
```
Comment (passsy) [inline lib/src/common/device_info/user_agent_parser.dart:13]: Shouldn't we better send the whole `userAgent` string to the server and parse it there?

Comment (brianegan-wiredash) [reply lib/src/common/device_info/user_agent_parser.dart:13]: 🤦 Yes. Refactor inc haha

**Pull request merged**